src/webapi/events/socket.rs

```diff
@@ -78,7 +78,7 @@ impl ConcreteEvent for SocketErrorEvent {
 // https://html.spec.whatwg.org/#event-open
 // https://html.spec.whatwg.org/multipage/web-sockets.html#handler-websocket-onopen
 #[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
-#[reference(instance_of = "Event")] // TODO: Better type check.
+#[reference(instance_of = "Event", constraint="type=open")]
```

Comment:
Hmm... now that I think about it I guess I'd really like if whatever is in constraint would be an actual JS expression (right now this supports only strings, and is not forward compatible with adding e.g. integers or other types), which means that in case of a string it would have to contain `"`, which means that we *probably* would want a different outer delimiter than `"`? `constraint=(type = "open")` maybe? (As long as it's reasonable anything should be fine with me.)

Or maybe make it standalone, e.g. `#[reference(constraint(type = "open"))]`? That would probably work better if we'd want to support multiple constraints (since the user would then just simply duplicate the `#[reference(constraint(...))]` line)

Reply:
@koute I'm not sure I entirely understand what you're suggesting. 

I weighed the pros and cons of making it a JS expression, but constraint="type='open' "  or constraint(type = " 'open' ") seemed like overkill, given the dearth of non-string scenarios. Also, I'm uncertain if constraint(type=123) is discernible from constraint(type="123"). As far as I can tell, there should be no breaking changes if we decide to change it later to address future scenarios.

Additionally, I was unable to immediately conceive of any directives other than instance_of for which the constraint attribute would be relevant.

With the above having been said, I'm happy to make whatever changes you feel are appropriate if you can help me understand which specific syntax you would prefer :)

Reply:
My point is not to support arbitrary JS expressions right now (we can ignore other types and support only strings for now), but to have a forward compatible syntax. (:

> Also, I'm uncertain if constraint(type=123) is discernible from constraint(type="123"). As far as I can tell, there should be no breaking changes if we decide to change it later to address future scenarios.

*Currently* as this PR stands `type=123` means that `type` field needs to be equal to `"123"` (that is - to a string). So, say, we'd want to extend this with support for integers - how will we be able to tell if `type=123` means `123` (an integer) or `"123"` (a string)?

Basically, have the most forward compatible syntax possible, but be as conservative as possible with it.

> Additionally, I was unable to immediately conceive of any directives other than instance_of for which the constraint attribute would be relevant.

How about having a constraint *without* an `instance_of`? (: I mean, we *probably* won't use it internally, but this is also usable by our downstream users, and they might find a use for it. (That is - make the object convertible based on a value of a field *without* discriminating based on what it's an instance of.)

And, I think that purely *syntax-wise* is should be cleaner to have the constraint in its own separate thing than shoehorned into the `instance_of`. Especially in case of multiple constraints, e.g.

```
#[reference(instance_of = "Event", constraint=..., constraint=..., constraint=...)]
```

vs

```
#[reference(constraint=...)]
#[reference(constraint=...)]
#[reference(constraint=...)]
```

It might not be the same for everyone, but for me *personally* the second one is more obvious in that it's actually *adding* constraints instead of overriding them.

That said, of course, I might be wrong, and if you disagree with some of my reasoning then I'm more than happy to discuss it.

Reply:
> If they were to put reference(contraint...) after the reference(instance_of) then they would get an undesirable outcome, insofar as the constraints would be silently ignored.

I don't understand - how so? The constraints would never be silently ignored, and it shouldn't matter whenever the user specifies the `constraint` or `instance_of` first - it should work exactly the same. *If* the `instance_of` is specified then that is always checked first, then every `constraint`s are checked in the order of their appearance. Nothing overrides anything - those are declarative attributes, not imperative commands. (This would apply equally in both the single-line case and the multiple-line case.)

> Additionally, I think it may make sense to change constraint to constraint_string and also add constraint_js for raw js. I don't anticipate we would need integer comparisons, but conceivably one could add constraint_int as well. Does that make sense?

But, you see, why not simply make it *look* like it's a string in the first place? (: I would rather avoid having multiple `constraint_foo`, `constraint_bar` keys, and I think it just makes more sense to have a single one and just figure out the type from the value. `"123"` is a string, `123` is a number, etc.

I do agree that something like `constraint="foo=\"foo\""` or `constraint="foo='foo'"` looks super awkward; perhaps this is why you don't like the idea of putting a quoted value in there? In that case as I've previously suggested we can just use a different syntax and have the leading `="` and trailing `"` replaced by `(` and `)` respectively.